## Practice Exercises

### Importing libraries

In [2]:
import pandas as pd

In [2]:
import numpy as np

In [4]:
import os 

### Importing Dataframes

In [6]:
path = r"C:\Users\user\OneDrive - Universitat de Barcelona\Instacart Basket Analysis"

In [7]:
df_prods= pd.read_csv(os.path.join(path,"02 Data","Original Data", "products.csv"),index_col=False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

### Create a dataframe

In [7]:
df_test= pd.DataFrame()

### Create a mixed type column

In [8]:
df_test["mix"]= ["a","b", 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

### The function for checking whether a dataframe contains any mixed-type columns is as follows

In [10]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


### Changing to only one data type

In [11]:
df_test['mix'] = df_test['mix'].astype('str')

### To fix missing values, you first need to be able to find them. Let’s walk through how to do so in your df_prods dataframe

In [12]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

### To actually view these 16 values, you can create a subset of the dataframe containing only the values in question

In [13]:
df_nan= df_prods[df_prods["product_name"].isnull()==True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

### Filter out the ones that aren’t missing into a subset dataframe and continue your analysis with this new dataframe.

In [16]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
df_prods_clean.shape

(49677, 5)

### Another way you can drop all missing values is via the following command:

In [18]:
df_prods.dropna(inplace = True)

### If you wanted to use this command to drop only the NaNs from a particular column

In [19]:
df_prods.dropna(subset = ["product_name"], inplace = True)

### The following command will look for full duplicates within your dataframe:

In [20]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49677, 5)

### create a new dataframe that doesn’t include the duplicates you just identified using the drop_duplicates() function

In [23]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

### export your new dataframe as a .”csv” file and store it in your “Prepared Data” folder

In [25]:
df_prods_clean_no_dups.to_csv(os.path.join(path,"02 Data","Prepared Data","products_checked.csv"))

## Task exercises

In [26]:
# Run the df.describe() function on your df_ords dataframe. 
# Share in a markdown cell whether anything about the data looks off or should be investigated further

In [27]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

**The data looks fine overall. The "Unnamed: 0" column may be unnecessary. order_number max is 100, which seems high but could be valid. days_since_prior_order maxes at 30, possibly a data cap worth checking. Everything else looks normal.**

In [28]:
# Check for mixed-type data in your df_ords dataframe

In [29]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [30]:
# If you find mixed-type data, fix it. The column in question should contain observations of a single data type

**No columns with a mixed-type data were found.**

In [31]:
# Run a check for missing values in your df_ords dataframe.
# In a markdown cell, report your findings and propose an explanation for any missing values you find

In [32]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

**The dataset is mostly complete, except for 206,209 missing values in days_since_prior_order. These are likely from first-time orders, where no previous order exists to calculate the gap**

In [33]:
# Address the missing values using an appropriate method
# In a markdown cell, explain why you used your method of choice

**I chose to to leave it null as is**

In [37]:
# Run a check for duplicate values in your df_ords data
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find

In [44]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [39]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

**There are no duplicate rows**

In [40]:
# Address the duplicates using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

**There are no duplicate rows**

In [41]:
# Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names

In [42]:
df_prods_clean_no_dups.to_csv(os.path.join(path,"02 Data","Prepared Data","products_checked.csv"))

In [45]:
df_ords.to_csv(os.path.join(path,"02 Data","Prepared Data","order_checked.csv"))